File: demos/01-cli-agent/src/providers/zhipuai.ts
import type { LLMClient, ChatParams, ChatResponse } from './base.js';
import { logger, formatJSON } from '../logger.js';

/**
 * 智谱 AI 客户端适配器
 * 支持普通 API 和 CodePlan API
 */
export class ZhipuClient implements LLMClient {
  private apiKey: string;
  private baseURL: string;

  constructor(apiKey: string, baseURL?: string) {
    this.apiKey = apiKey;
    // 支持 CodePlan 专用端点或普通端点
    this.baseURL = baseURL || 'https://open.bigmodel.cn/api/paas/v4';
  }

  async chat(params: ChatParams): Promise<ChatResponse> {
    const url = `${this.baseURL}/chat/completions`;

    logger.debug('ZHIPUAI', `📡 请求端点: ${url}`);
    logger.debug('ZHIPUAI', `🤖 模型: ${params.model}`);

    // 转换消息格式（智谱 OpenAI 兼容格式）
    const messages = params.messages.map(m => {
      const msg: any = {
        role: m.role === 'function' ? 'user' : m.role,
        content: m.content
      };

      // 如果是工具调用结果，添加名称
      if (m.role === 'function') {
        msg.name = (m as any).name;
        msg.content = `工具 ${(m as any).name} 的执行结果：${m.content}`;
      }

      return msg;
    });

    logger.debug('ZHIPUAI', `📤 消息数量: ${messages.length}`);

    const requestBody = {
      model: params.model,
      messages: messages,
      stream: params.stream ?? false
    };

    logger.debug('ZHIPUAI', `📦 请求体: ${formatJSON(requestBody, 0)}`);

    try {
      const startTime = Date.now();
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`
        },
        body: JSON.stringify(requestBody)
      });

      const duration = Date.now() - startTime;
      logger.debug('ZHIPUAI', `⏱️  响应时间: ${duration}ms`);

      if (!response.ok) {
        const errorText = await response.text();
        logger.error('ZHIPUAI', `❌ HTTP ${response.status}: ${errorText}`);
        throw new Error(`HTTP ${response.status}: ${errorText}`);
      }

      const data: any = await response.json();

      // 检查 API 错误
      if (data.error) {
        logger.error('ZHIPUAI', `❌ API 错误 [${data.error.code}]: ${data.error.message}`);
        throw new Error(`智谱 AI 错误 [${data.error.code}]: ${data.error.message}`);
      }

      // 提取响应内容
      const content = data.choices?.[0]?.message?.content || '';
      logger.debug('ZHIPUAI', `📥 响应内容: ${content.substring(0, 100)}...`);

      // 提取 Token 使用
      const usage = data.usage ? {
        prompt_tokens: data.usage.prompt_tokens,
        completion_tokens: data.usage.completion_tokens,
        total_tokens: data.usage.total_tokens
      } : undefined;

      logger.debug('ZHIPUAI', `📊 Token: ${formatJSON(usage)}`);

      return { content, usage };
    } catch (error: any) {
      logger.error('ZHIPUAI', `💥 调用异常: ${error.message}`);
      throw new Error(`智谱 AI 调用失败: ${error.message}`);
    }
  }
}

File: demos/03-memory-agent/backend/app/providers/zhipuai.py
"""
智谱 AI 客户端适配器
支持 Function Calling API 和流式输出
"""
import json
import logging
import httpx
from typing import List, Dict, Any, AsyncIterator

from app.providers.base import LLMClient, ChatMessage, ChatResponse, ToolCall
logger = logging.getLogger(__name__)

class ZhipuClient(LLMClient):
    """智谱 AI 客户端"""

    def __init__(self, api_key: str, base_url: str = "https://open.bigmodel.cn/api/coding/paas/v4"):
        self.api_key = api_key
        self.base_url = base_url

    async def chat(
        self,
        model: str,
        messages: List[ChatMessage],
        tools: List[Dict[str, Any]] | None = None,
        stream: bool = False
    ) -> ChatResponse:
        """发送聊天请求"""
        url = f"{self.base_url}/chat/completions"

        # 转换消息格式
        request_messages = [self._convert_message(m) for m in messages]

        logger.info(f"[智普 chat] request_messages: {request_messages}")
        # 构建请求体
        request_body = {
            "model": model,
            "messages": request_messages,
            "stream": False
        }

        if tools:
            request_body["tools"] = tools

        async with httpx.AsyncClient() as client:
            response = await client.post(
                url,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}"
                },
                json=request_body,
                timeout=60.0
            )

            if response.status_code != 200:
                raise Exception(f"API 错误: {response.status_code} - {response.text}")

            data = response.json()

            if "error" in data:
                raise Exception(f"智谱 AI 错误: {data['error']}")

            # 提取响应
            choice = data.get("choices", [{}])[0]
            message = choice.get("message", {})

            # 处理 content（GLM-5 可能同时返回 reasoning_content 和 content）
            # reasoning_content 是思维链，content 是实际回复内容
            # 对于摘要生成等场景，应该使用 content 而不是 reasoning_content
            content = message.get("content") or message.get("reasoning_content") or ""
            tool_calls = []

            # 提取工具调用
            if "tool_calls" in message and message["tool_calls"]:
                for tc in message["tool_calls"]:
                    func_args = tc["function"]["arguments"]
                    if isinstance(func_args, str):
                        func_args = json.loads(func_args)

                    tool_calls.append(ToolCall(
                        id=tc["id"],
                        type="function",
                        function={
                            "name": tc["function"]["name"],
                            "arguments": func_args
                        }
                    ))

            # 提取使用情况
            usage = None
            if "usage" in data:
                usage = {
                    "prompt_tokens": data["usage"]["prompt_tokens"],
                    "completion_tokens": data["usage"]["completion_tokens"],
                    "total_tokens": data["usage"]["total_tokens"]
                }

            return ChatResponse(
                content=content,
                tool_calls=tool_calls,
                usage=usage
            )

    async def chat_stream(
        self,
        model: str,
        messages: List[ChatMessage],
        tools: List[Dict[str, Any]] | None = None
    ) -> AsyncIterator[Dict[str, Any]]:
        """
        发送流式聊天请求

        返回事件字典:
        - {"event": "reasoning", "content": "..."} - 思维链内容
        - {"event": "content", "content": "..."} - 普通内容
        - {"event": "tool_call", "tool_calls": [...]} - 工具调用
        """
        url = f"{self.base_url}/chat/completions"

        # 转换消息格式
        request_messages = [self._convert_message(m) for m in messages]
        logger.info(f"[智普zhipu chat] request_messages: {json.dumps(request_messages, ensure_ascii=False)}")

        # 构建请求体
        request_body = {
            "model": model,
            "messages": request_messages,
            "stream": True
        }

        # 启用思维链功能（仅 GLM-4.5+ 和 GLM-4.1V-Thinking 支持）
        if model.startswith("glm-4.5") or "thinking" in model.lower():
            request_body["thinking"] = {"type": "enabled"}

        if tools:
            request_body["tools"] = tools

        logger.info(f"[智普zhipu chat_stream] request_body: {json.dumps(request_body, ensure_ascii=False, indent=2)}")

        # 用于累积 tool_calls 数据
        current_tool_calls: List[Dict[str, Any]] = []

        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST",
                url,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}"
                },
                json=request_body,
                timeout=60.0
            ) as response:
                if response.status_code != 200:
                    error_text = await response.aread()
                    error_text = error_text.decode('utf-8', errors='ignore')
                    raise Exception(f"API 错误: {response.status_code} - {error_text}")

                async for line in response.aiter_lines():
                    if not line or not line.startswith("data: "):
                        continue

                    data_str = line[6:]  # 移除 "data: " 前缀

                    if data_str == "[DONE]":
                        break

                    try:
                        data = json.loads(data_str)
                        choices = data.get("choices", [])
                        if choices:
                            delta = choices[0].get("delta", {})
                            finish_reason = choices[0].get("finish_reason", "")

                            # 检测 tool_calls（流式模式）
                            if "tool_calls" in delta and delta["tool_calls"]:
                                # 累积 tool_calls 数据
                                for tc in delta["tool_calls"]:
                                    index = tc.get("index", 0)
                                    # 确保数组有足够空间
                                    while len(current_tool_calls) <= index:
                                        current_tool_calls.append({
                                            "id": "",
                                            "type": "function",
                                            "function": {"name": "", "arguments": ""}
                                        })

                                    # 更新 tool_call 数据
                                    if "id" in tc:
                                        current_tool_calls[index]["id"] = tc["id"]
                                    if "function" in tc:
                                        func = tc["function"]
                                        if "name" in func:
                                            current_tool_calls[index]["function"]["name"] = func["name"]
                                        if "arguments" in func:
                                            # arguments 是逐步追加的
                                            current_tool_calls[index]["function"]["arguments"] += func["arguments"]

                            # 检测是否完成（finish_reason = tool_calls）
                            if finish_reason == "tool_calls" and current_tool_calls:
                                logger.info(f"[流式工具调用检测] tool_calls: {current_tool_calls}")
                                yield {
                                    "event": "tool_call",
                                    "tool_calls": current_tool_calls
                                }
                                current_tool_calls = []

                            # 处理普通内容（reasoning_content 和 content）
                            # GLM-4.5+ 使用 reasoning_content 字段表示思维链
                            # GLM-4 使用 content 字段表示普通内容
                            reasoning_content = delta.get("reasoning_content", "")
                            content = delta.get("content", "")

                            if reasoning_content:
                                yield {
                                    "event": "reasoning",
                                    "content": reasoning_content
                                }
                            elif content:
                                yield {
                                    "event": "content",
                                    "content": content
                                }
                    except (json.JSONDecodeError, KeyError):
                        continue

    def _convert_message(self, message: ChatMessage) -> Dict[str, Any]:
        """转换消息格式"""
        return {
            "role": message.role,
            "content": message.content
        }
